lib/veri_policy.ml

```diff
@@ -0,0 +1,171 @@
+  type t = g * g
+  module V = struct
+    type t = Source | Sink | Person of int | Task of int [@@deriving compare]
+    let hash = Hashtbl.hash
+    let equal x y = compare x y = 0
+  end
+  module E = struct
+    type label = unit
+
+    type t = {src : V.t; dst : V.t} [@@deriving fields]
+    let make src dst = {src; dst}
+    let label _ = ()
+  end
+  type dir = Succ | Pred
+
+  let iter dir f ((expect, workers), (expect', jobs)) v = 
+    match v,dir with
+    | V.Source,Pred -> ()
+    | V.Source,Succ ->
+      Array.iteri workers ~f:(fun i _  ->
+          f @@ E.make V.Source (V.Person i))
+    | V.Sink,Pred ->
+      Array.iteri jobs ~f:(fun i _ ->
+          f @@ E.make (V.Person i) V.Sink)
+    | V.Sink,Succ -> ()
+    | V.Person i as p,Pred ->
+      f @@ E.make V.Source p
+    | V.Person i as p,Succ ->
+      Array.iteri jobs ~f:(fun j job ->
+          if sat_events (expect,workers.(i)) (expect', job) then 
+            f (E.make p (V.Task j)))
+    | V.Task j as t,Succ ->
+      f @@ E.make t V.Sink
+    | V.Task j as t,Pred ->
+      Array.iteri workers ~f:(fun i worker ->
+          if sat_events (expect, worker) (expect', jobs.(j)) then 
+            f (E.make (V.Person i) t))
+
+  let iter_succ_e = iter Succ
+  let iter_pred_e = iter Pred
+end
+
+module F = struct
+  type t = int
+  type label = unit
+  let max_capacity () = 1
+  let min_capacity () = 0
+  let flow () = min_capacity ()
+  let add = (+)
+  let sub = (-)
+  let zero = 0
+  let compare = Int.compare
+end
+
+module FFMF = Flow.Ford_Fulkerson(G)(F)
+
+let single_match trial events =
+  List.filter ~f:(sat_event trial) (Set.to_list events)
+
+let is_empty_insn  t = Field.is_empty t.insn.field
+let is_empty_left  t = Field.is_empty t.left.field
+let is_empty_right t = Field.is_empty t.right.field
+let is_sat_insn t insn = not (is_empty_insn t) && sat t.insn insn
+
+let match_events t insn events events' =
+  match is_sat_insn t insn with
+  | false -> []
+  | true -> 
+    match is_empty_left t, is_empty_right t with
+    | true, _ ->
+      single_match t.right events' |>
+      List.map ~f:(fun e -> None, Some e)
+    | _, true -> 
+      single_match t.left events |>
+      List.map ~f:(fun e -> Some e, None)
+    | _ -> 
+      let workers = t.left, Set.to_array events in      
+      let jobs = t.right, Set.to_array events' in
+      let (flow,_) = FFMF.maxflow (workers, jobs) G.V.Source G.V.Sink  in
+      Array.foldi (snd workers) ~init:[] 
+        ~f:(fun i acc w ->   
+            match Array.findi (snd jobs) ~f:(fun j e ->
+                flow (G.E.make (G.V.Person i) (G.V.Task j)) <> 0) with
+            | None -> acc 
+            | Some (_,e) -> (Some w, Some e) :: acc)  
+
+let remove what from = 
+  List.fold_left ~init:from ~f:Set.remove what
+
+let process t insn events events' : r = 
+  match match_events t insn events events' with 
+  | [] -> Ok (events, events')
+  | res ->
+    let evs, evs' = List.fold_left ~init:([],[]) res
```

Comment:
There is a `List.partition` function for this. Also, you're performing too much intermediate steps, for example, why to partition a list of matches into two lists, and then remove each list from the corresponding set, if you can just apply this action as you fold the list. Moreover, why to create the list of pairs on the first hand, if you can just parametrize `match_events` with the action (of course in this case a better name for the function would be `fold_matches`).